File: cpp/array_search/hash_search_production.cpp
#include <cstddef>
#include <optional>
#include <unordered_map>
#include <vector>

class HashIndex {
public:
    explicit HashIndex(const std::vector<int>& values) {
        index_.reserve(values.size());

        for (std::size_t position = 0; position < values.size(); ++position) {
            index_.try_emplace(values[position], position);
        }
    }

    std::optional<std::size_t> find(int target) const {
        const auto it = index_.find(target);

        if (it == index_.end()) {
            return std::nullopt;
        }

        return it->second;
    }

private:
    std::unordered_map<int, std::size_t> index_;
};

int hash_search(const std::vector<int>& values, int target) {
    HashIndex index(values);
    const auto result = index.find(target);
    return result ? static_cast<int>(*result) : -1;
}

File: python/array_search/hash_search_production.py
class HashIndex: 
    def __init__(self, values):
        self._index = {}

        for position, value in enumerate(values):
            self._index.setdefault(value, position)

    def find(self, target):
        return self._index.get(target, -1)

    def contains(self, target):
        return target in self._index


def hash_search(values, target):
    return HashIndex(values).find(target)
